File: code/richediting/src/main/java/org/fbme/ide/richediting/inspections/NetworkInspectionsFacility.java
package org.fbme.ide.richediting.inspections;

import jetbrains.mps.editor.runtime.style.StyleAttributes;
import jetbrains.mps.editor.runtime.style.StyleImpl;
import jetbrains.mps.nodeEditor.EditorSettings;
import jetbrains.mps.nodeEditor.MPSColors;
import jetbrains.mps.nodeEditor.MPSFonts;
import jetbrains.mps.nodeEditor.cells.TextLine;
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionController;
import org.fbme.ide.richediting.adapters.fbnetwork.FBNetworkComponentController;
import org.fbme.ide.richediting.viewmodel.NetworkComponentView;
import org.fbme.ide.richediting.viewmodel.NetworkConnectionView;
import org.fbme.ide.richediting.viewmodel.NetworkPortView;
import org.fbme.ide.richediting.viewmodel.NetworkView;
import org.fbme.lib.common.Declaration;
import org.fbme.lib.common.Element;
import org.fbme.lib.iec61499.fbnetwork.FBNetworkConnection;
import org.fbme.lib.iec61499.fbnetwork.FunctionBlockDeclarationBase;
import org.fbme.lib.iec61499.fbnetwork.PortPath;
import org.fbme.lib.iec61499.instances.NetworkInstance;
import org.fbme.scenes.controllers.ROLayoutModel;
import org.fbme.scenes.controllers.scene.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

public class NetworkInspectionsFacility {

    private static final int PORT_TEXT_X_OFFSET = 5;
    private static final int PORT_TEXT_Y_OFFSET = 2;
    private static final int COMPONENT_TEXT_Y_OFFSET = 10;

    @NotNull
    private final NetworkInstance myInstance;

    @NotNull
    private final SceneEditor myEditor;

    @NotNull
    private final Layer myLayer;

    @NotNull
    private final Function<NetworkComponentView, FBNetworkComponentController> myComponentProvider;

    @NotNull
    private final Function<NetworkConnectionView, FBConnectionController> myConnectionProvider;

    @NotNull
    private final ROLayoutModel<NetworkComponentView> myLayoutModel;

    private final HashMap<PortPath<? extends Declaration>, InspectionData> myPortInspections = new HashMap<>();
    private final HashMap<Declaration, InspectionData> myComponentInspections = new HashMap<>();
    private final NetworkView myNetworkView;

    private final MyLayouter myLayouter = new MyLayouter();
    private final MyPainter myPainter = new MyPainter();

    private final Set<FBNetworkConnection> myHighlightedConnections = new HashSet<>();


    public NetworkInspectionsFacility(@NotNull NetworkView networkView, @NotNull NetworkInstance networkInstance, @NotNull SceneEditor editor, @NotNull Function<NetworkComponentView, FBNetworkComponentController> componentProvider, @NotNull Function<NetworkConnectionView, FBConnectionController> connectionProvider, @NotNull ROLayoutModel<NetworkComponentView> layoutModel, @NotNull Layer layer) {
        myInstance = networkInstance;
        myNetworkView = networkView;
        myEditor = editor;
        myComponentProvider = componentProvider;
        myConnectionProvider = connectionProvider;
        myLayoutModel = layoutModel;
        myLayer = layer;
        layoutModel.addListener(new ROLayoutModel.Listener<>() {
            public void onComponentAdded(@NotNull NetworkComponentView component, int x, int y) {
                // do nothing
            }

            public void onComponentRemoved(@NotNull NetworkComponentView component) {
                // do nothing
            }

            public void onComponentsMoved(@NotNull Set<NetworkComponentView> components, int dx, int dy, boolean completed) {
                myLayouter.relayout();
            }
        });
        myEditor.addInitializer(new MyInitializer());
    }

    public void setInspectionForPort(@NotNull PortPath<?> port, @Nullable Inspection inspection) {
        if (inspection == null) {
            myPortInspections.remove(port);
        } else {
            InspectionData data = new InspectionData(inspection);
            relayoutPortInspection(port, data);
            myPortInspections.put(port, data);
        }
        myEditor.fireRepaint();
    }

    public void setInspectionForComponent(@NotNull Declaration component, @Nullable Inspection inspection) {
        if (inspection == null) {
            myComponentInspections.remove(component);
        } else {
            InspectionData data = new InspectionData(inspection);
            relayoutComponentInspection(component, data);
            myComponentInspections.put(component, data);
        }
        myEditor.fireRepaint();
    }

    public void setHighlightingForConnection(@NotNull FBNetworkConnection connection, @Nullable Color color) {
        FBConnectionController controller = myConnectionProvider.apply(myNetworkView.getConnectionView(connection));
        if (color != null) {
            myHighlightedConnections.add(connection);
        } else {
            myHighlightedConnections.remove(connection);
        }
        controller.setHighlightColor(color);
        myEditor.fireRepaint();
    }

    public void clear() {
        myPortInspections.clear();
        myComponentInspections.clear();
        for (FBNetworkConnection connection : myHighlightedConnections) {
            FBConnectionController controller = myConnectionProvider.apply(myNetworkView.getConnectionView(connection));
            controller.setHighlightColor(null);
        }
        myEditor.fireRepaint();
    }

    public void install() {
        myEditor.addPainter(myLayer, myPainter);
        myEditor.addLayouter(myLayouter);
    }

    public void uninstall() {
        myEditor.removeLayouter(myLayouter);
        myEditor.removePainter(myPainter);
    }

    private static class InspectionData {
        private final Inspection myInspection;

        private final Color myColor;

        private int myX;

        private int myY;

        private final TextLine myTextline;

        private InspectionData(Inspection inspection) {
            myInspection = inspection;

            myColor = inspection.getColor();
            StyleImpl style = new StyleImpl();
            style.set(StyleAttributes.FONT_SIZE, EditorSettings.getInstance().getFontSize() * 3 / 4);
            style.set(StyleAttributes.TEXT_COLOR, myColor);
            style.set(StyleAttributes.FONT_STYLE, MPSFonts.BOLD);
            myTextline = new TextLine(inspection.getText(), style, false);
        }

        private void paint(Graphics g) {
            g.setColor(MPSColors.WHITE);
            g.fillRect(myX - 1, myY, myTextline.getWidth() + 2, myTextline.getHeight());
            g.setColor(myColor);
            g.drawRect(myX - 1, myY, myTextline.getWidth() + 2, myTextline.getHeight());
            myTextline.paint(g, myX, myY);
        }
    }

    private class MyLayouter implements SceneLayouter {

        @Override
        public void relayout() {
            for (Map.Entry<Declaration, InspectionData> componentInspection : myComponentInspections.entrySet()) {
                Declaration key = componentInspection.getKey();
                InspectionData data = componentInspection.getValue();
                relayoutComponentInspection(key, data);
            }
            for (Map.Entry<PortPath<? extends Declaration>, InspectionData> portInspection : myPortInspections.entrySet()) {
                PortPath<? extends Declaration> key = portInspection.getKey();
                InspectionData data = portInspection.getValue();
                relayoutPortInspection(key, data);
            }
        }

        @NotNull
        @Override
        public Rectangle getLayoutBounds() {
            return new Rectangle(-1, -1);
        }
    }

    private class MyPainter implements ScenePainter {

        @Override
        public void paint(Graphics2D graphics) {
            for (InspectionData data : myComponentInspections.values()) {
                data.paint(graphics);
            }
            for (InspectionData data : myPortInspections.values()) {
                data.paint(graphics);
            }
        }
    }

    private class MyInitializer implements SceneInitializer {

        @Override
        public void onAdd() {
            NetworkInspectionsData inspectionsData = myEditor.loadState(NetworkInspectionsData.KEY);
            if (inspectionsData != null) {
                loadState(inspectionsData);
            }
            InspectionManagerImpl.getInstance(myEditor.getContext().getEditorComponent()).registerNetwork(myInstance, NetworkInspectionsFacility.this);
        }

        @Override
        public void onRemove() {
            NetworkInspectionsData data = new NetworkInspectionsData();
            storeState(data);
            myEditor.storeState(NetworkInspectionsData.KEY, data);
            InspectionManager manager = InspectionManagerImpl.getInstance(myEditor.getContext().getEditorComponent());
            if (manager != null) {
                manager.unregisterNetwork(myInstance);
            }
        }
    }

    private void loadState(NetworkInspectionsData data) {
        for (Map.Entry<Declaration, Inspection> componentInspection : data.componentInspections.entrySet()) {
            myComponentInspections.put(componentInspection.getKey(), new InspectionData(componentInspection.getValue()));
        }
        for (Map.Entry<PortPath<? extends Declaration>, Inspection> portInspection : data.portInspections.entrySet()) {
            myPortInspections.put(portInspection.getKey(), new InspectionData(portInspection.getValue()));
        }
        for (Map.Entry<FBNetworkConnection, Color> connection : data.highlightedConnections.entrySet()) {
            myHighlightedConnections.add(connection.getKey());
            FBConnectionController controller = myConnectionProvider.apply(myNetworkView.getConnectionView(connection.getKey()));
            controller.setHighlightColor(connection.getValue());
        }
        if (!data.highlightedConnections.isEmpty()) {
            myEditor.fireRepaint();
        }
    }

    private void storeState(NetworkInspectionsData data) {
        for (Map.Entry<Declaration, InspectionData> componentInspection : myComponentInspections.entrySet()) {
            data.componentInspections.put(componentInspection.getKey(), componentInspection.getValue().myInspection);
        }
        for (Map.Entry<PortPath<? extends Declaration>, InspectionData> portInspection : myPortInspections.entrySet()) {
            data.portInspections.put(portInspection.getKey(), portInspection.getValue().myInspection);
        }
        for (FBNetworkConnection connection : myHighlightedConnections) {
            FBConnectionController controller = myConnectionProvider.apply(myNetworkView.getConnectionView(connection));
            data.highlightedConnections.put(connection, controller.getHighlightColor());
        }
    }

    private void relayoutComponentInspection(Element key, InspectionData data) {
        NetworkComponentView view = myNetworkView.getComponentView(key);
        FBNetworkComponentController controller = myComponentProvider.apply(view);
        Rectangle bounds = controller.getBounds(myLayoutModel.getActivePosition(view));
        data.myTextline.relayout();
        data.myY = bounds.y + bounds.height + COMPONENT_TEXT_Y_OFFSET + data.myTextline.getHeight();
        data.myX = bounds.x + bounds.width / 2 - data.myTextline.getWidth() / 2;
    }

    private void relayoutPortInspection(PortPath<?> key, InspectionData data) {
        FunctionBlockDeclarationBase component = key.getFunctionBlock();
        NetworkComponentView view = myNetworkView.getComponentView(component);
        NetworkPortView portView = myNetworkView.getPortView(key);
        FBNetworkComponentController controller = myComponentProvider.apply(view);
        Point coordinates = controller.getPortCoordinates(portView, myLayoutModel.getActivePosition(view));
        data.myTextline.relayout();
        data.myY = coordinates.y - PORT_TEXT_Y_OFFSET - data.myTextline.getHeight() / 2;
        data.myX = coordinates.x + (controller.isSource(portView) ? PORT_TEXT_X_OFFSET : -(data.myTextline.getWidth() + PORT_TEXT_X_OFFSET));
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/inspections/NetworkInspectionsFacility.kt
package org.fbme.ide.richediting.inspections

import jetbrains.mps.editor.runtime.style.Padding
import jetbrains.mps.editor.runtime.style.StyleAttributes
import jetbrains.mps.nodeEditor.EditorSettings
import jetbrains.mps.nodeEditor.MPSColors
import jetbrains.mps.nodeEditor.MPSFonts
import jetbrains.mps.nodeEditor.cells.EditorCell_Basic
import jetbrains.mps.nodeEditor.cells.ParentSettings
import jetbrains.mps.nodeEditor.cells.TextLine
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionController
import org.fbme.ide.richediting.adapters.fbnetwork.FBNetworkComponentController
import org.fbme.ide.richediting.adapters.fbnetwork.fb.AbstractFBCell.Companion.PORT_SIZE
import org.fbme.ide.richediting.viewmodel.NetworkComponentView
import org.fbme.ide.richediting.viewmodel.NetworkConnectionView
import org.fbme.ide.richediting.viewmodel.NetworkView
import org.fbme.lib.common.Declaration
import org.fbme.lib.common.Element
import org.fbme.lib.iec61499.fbnetwork.FBNetworkConnection
import org.fbme.lib.iec61499.fbnetwork.PortPath
import org.fbme.lib.iec61499.instances.NetworkInstance
import org.fbme.scenes.controllers.LayoutUtil
import org.fbme.scenes.controllers.ROLayoutModel
import org.fbme.scenes.controllers.components.ComponentController
import org.fbme.scenes.controllers.scene.*
import java.awt.*

class NetworkInspectionsFacility(
    private val networkView: NetworkView,
    private val instance: NetworkInstance,
    private val editor: SceneEditor,
    val componentProvider: (NetworkComponentView) -> FBNetworkComponentController,
    private val connectionProvider: (NetworkConnectionView) -> FBConnectionController,
    val layoutModel: ROLayoutModel<NetworkComponentView>,
    private val layer: Layer
) {
    private val portInspections = HashMap<PortPath<out Declaration>?, InspectionData>()
    private val componentInspections = HashMap<Declaration, InspectionData>()
    private val layouter: MyLayouter = MyLayouter()
    private val painter: MyPainter = MyPainter()
    private val highlightedConnections: MutableSet<FBNetworkConnection> = HashSet()

    fun setInspectionForPort(port: PortPath<*>, inspection: Inspection?) {
        if (inspection == null) {
            portInspections.remove(port)
        } else {
            val data = InspectionData(inspection)
            relayoutPortInspection(port, data)
            portInspections[port] = data
        }
        editor.fireRepaint()
    }

    fun setInspectionForComponent(component: Declaration, inspection: Inspection?) {
        if (inspection == null) {
            componentInspections.remove(component)
        } else {
            val data = InspectionData(inspection)
            relayoutComponentInspection(component, data)
            componentInspections[component] = data
        }
        editor.fireRepaint()
    }

    fun setHighlightingForConnection(connection: FBNetworkConnection, color: Color?) {
        val controller = connectionProvider(networkView.getConnectionView(connection))
        if (color != null) {
            highlightedConnections.add(connection)
        } else {
            highlightedConnections.remove(connection)
        }
        controller.highlightColor = color
        editor.fireRepaint()
    }

    fun clear() {
        portInspections.clear()
        componentInspections.clear()
        for (connection in highlightedConnections) {
            val controller = connectionProvider(networkView.getConnectionView(connection))
            controller.highlightColor = null
        }
        editor.fireRepaint()
    }

    fun install() {
        editor.addPainter(layer, painter)
        editor.addLayouter(layouter)
    }

    fun uninstall() {
        editor.removeLayouter(layouter)
        editor.removePainter(painter)
    }

    private inner class InspectionData(val inspection: Inspection) : EditorCell_Basic(editor.editorContext, null) {
        private val textLine: TextLine
        private var lineSize: Int = 0

        init {
            setTextColor(inspection.color)
            setFontStyle(if (inspection.bold) MPSFonts.BOLD else MPSFonts.PLAIN)
            textLine = TextLine(inspection.text, style, false, editor.editorComponentSettings)
        }

        override fun getDescent(): Int {
            return textLine.descent
        }

        override fun relayoutImpl() {
            textLine.relayout()
            myWidth = textLine.width
        }

        override fun paintContent(g: Graphics, parentSettings: ParentSettings?) {
            g.color = MPSColors.WHITE
            g.fillRect(x, y, width, lineSize)
            g.color = inspection.color
            g.drawRect(x, y, width, lineSize)
            textLine.paint(g, x, y - lineSize / 4)
        }

        fun setFontSize(fontSize: Int) {
            style.set(StyleAttributes.FONT_SIZE, fontSize)
        }

        fun setTextColor(color: Color) {
            style.set(StyleAttributes.TEXT_COLOR, color)
        }

        fun setFontStyle(fontStyle: Int) {
            style.set(StyleAttributes.FONT_STYLE, fontStyle)
        }

        fun setLineSize(lineSize: Int) {
            this@InspectionData.lineSize = lineSize
        }
    }

    private inner class MyLayouter : SceneLayouter {
        override fun relayout() {
            for ((key, data) in componentInspections) {
                relayoutComponentInspection(key, data)
            }
            for ((key, data) in portInspections) {
                relayoutPortInspection(key, data)
            }
        }

        override val layoutBounds: Rectangle
            get() = Rectangle(-1, -1)
    }

    private inner class MyPainter : ScenePainter {
        override fun paint(graphics: Graphics2D) {
            for (data in componentInspections.values) {
                data.paint(graphics)
            }
            for (data in portInspections.values) {
                data.paint(graphics)
            }
        }
    }

    private inner class MyInitializer : SceneInitializer {
        override fun onAdd() {
            val inspectionsData = editor.loadState(
                NetworkInspectionsData.KEY
            )
            inspectionsData?.let { loadState(it) }
            val manager = InspectionManagerImpl.getInstance(editor.editorContext.editorComponent)
            manager?.registerNetwork(instance, this@NetworkInspectionsFacility)
        }

        override fun onRemove() {
            val data = NetworkInspectionsData()
            storeState(data)
            editor.storeState(NetworkInspectionsData.KEY, data)
            val manager = InspectionManagerImpl.getInstance(editor.editorContext.editorComponent)
            manager?.unregisterNetwork(instance)
        }
    }

    private fun loadState(data: NetworkInspectionsData) {
        for ((key, value) in data.componentInspections) {
            componentInspections[key] = InspectionData(value)
        }
        for ((key, value) in data.portInspections) {
            portInspections[key] = InspectionData(
                value
            )
        }
        for ((key, value) in data.highlightedConnections) {
            highlightedConnections.add(key)
            val controller = connectionProvider(networkView.getConnectionView(key))
            controller.highlightColor = value
        }
        if (data.highlightedConnections.isNotEmpty()) {
            editor.fireRepaint()
        }
    }

    private fun storeState(data: NetworkInspectionsData) {
        for ((key, value) in componentInspections) {
            data.componentInspections[key] = value.inspection
        }
        for ((key, value) in portInspections) {
            data.portInspections[key!!] = value.inspection
        }
        for (connection in highlightedConnections) {
            val controller = connectionProvider(networkView.getConnectionView(connection))
            data.highlightedConnections[connection] = controller.highlightColor
        }
    }

    private fun relayoutComponentInspection(key: Element, data: InspectionData) {
        val view = networkView.getComponentView(key)
        val controller = componentProvider(view)
        val bounds = controller.getBounds(layoutModel.getActivePosition(view))
        data.setFontSize(LayoutUtil.getFontSize((controller as ComponentController<*>).componentCell.style))
        data.setLineSize(LayoutUtil.getLineSize((controller as ComponentController<*>).componentCell.style))
        data.requestRelayout()
        data.relayout()
        data.y = bounds.y + bounds.height + data.height
        data.x = bounds.x + bounds.width / 2 - data.width / 2
    }

    private fun relayoutPortInspection(key: PortPath<*>?, data: InspectionData) {
        val component = key!!.functionBlock ?: error("Can't get function block")
        val view = networkView.getComponentView(component)
        val portView = networkView.getPortView(key)
        val controller = componentProvider(view)
        val position = layoutModel.getActivePosition(view)
        val portBounds = controller.getPortBounds(portView, position)

        data.setFontSize(LayoutUtil.getFontSize((controller as ComponentController<*>).componentCell.style))
        data.setLineSize(LayoutUtil.getLineSize((controller as ComponentController<*>).componentCell.style))
        data.requestRelayout()
        data.relayout()
        data.x = if (controller.isSource(portView)) {
            portBounds.x + portBounds.width + scale(PORT_SIZE)
        } else {
            portBounds.x - scale(PORT_SIZE) - data.width
        }
        data.y = portBounds.y
    }

    private fun scale(size: Int): Int {
        return size * LayoutUtil.getFontSize(editor.editorContext.contextCell.style) / EditorSettings.getInstance().fontSize
    }

    init {
        layoutModel.addListener(object : ROLayoutModel.Listener<NetworkComponentView> {
            override fun onComponentAdded(component: NetworkComponentView, x: Int, y: Int) {
                // do nothing
            }

            override fun onComponentRemoved(component: NetworkComponentView) {
                // do nothing
            }

            override fun onComponentsMoved(
                components: Set<NetworkComponentView>,
                dx: Int,
                dy: Int,
                completed: Boolean
            ) {
                layouter.relayout()
            }
        })
        editor.addInitializer(MyInitializer())
    }
}